crates/cext/src/param.rs

```diff
@@ -0,0 +1,1120 @@
+/// [ParameterExpression].
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_free(expr: *mut Param) {
+    if !expr.is_null() {
+        if !expr.is_aligned() {
+            panic!("Attempted to free a non-aligned pointer.")
+        }
+
+        // SAFETY: We have verified the pointer is non-null and aligned, so it should be
+        // readable by Box.
+        unsafe {
+            let _ = Box::from_raw(expr);
+        }
+    }
+}
+
+/// @ingroup QkParam
+/// Construct a new Param from value
+///
+/// @param value real number for a new QkParam
+///
+/// @return A pointer to the created QkParam.
+///
+/// # Example
+///
+///     QkParam *r = qk_param_from_value(2.5);
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_param_from_double(value: f64) -> *mut Param {
+    let value = Param::Float(value);
+    Box::into_raw(Box::new(value))
+}
+
+/// @ingroup QkParam
+/// Construct a new QkParam from complex number
+///
+/// @param value complex number for a new QkParam
+///
+/// @return A pointer to the created QkParam.
+///
+/// # Example
+///
+///     QkParam *c = qk_param_from_complex((1.0,1.0));
+///
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub extern "C" fn qk_param_from_complex(value: Complex64) -> *mut Param {
+    let value = SymbolExpr::Value(Value::Complex(value));
+    let expr = ParameterExpression::from_symbol_expr(value);
+    let param = Param::ParameterExpression(Arc::new(expr));
+    Box::into_raw(Box::new(param))
+}
+
+// @ingroup QkParam
+// Construct a new QkParam from rational number
+//   TODO : this function needs PR #14686
+// @param numerator numerator of rational number
+// @param denominator denominator of rational number
+//
+// @return A pointer to the created QkParam.
+//
+// # Example
+//
+//     QkParam *c = qk_param_from_rational(1, 2);
+//
+//#[no_mangle]
+//#[cfg(feature = "cbinding")]
+//pub extern "C" fn qk_param_from_rational(numerator: i64, denominator: i64) -> *mut Param {
+//    let value = SymbolExpr::Value(Value::Rational{numerator, denominator});
+//    let expr = ParameterExpression::from_symbol_expr(value);
+//    let param = Param::ParameterExpression(Arc::new(expr));
+//    Box::into_raw(Box::new(param))
+//}
+
+/// @ingroup QkParam
+/// copy QkParam
+///
+/// @param name The name of symbol.
+///
+/// @return A pointer to the created QkParam.
+///
+/// # Example
+///
+///     QkParam *b = qk_param_copy(a);
+///
+/// # Safety
+///
+/// expr should be valid pointer to parameter to be copied
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_param_copy(expr: *const Param) -> *mut Param {
+    let expr = unsafe { const_ptr_as_ref(expr) };
+    Box::into_raw(Box::new(expr.clone()))
+}
+
+/// @ingroup QkParam
+/// get string expression of a Param.
+///
+/// @param expr A pointer to the Param
+///
+/// # Example
+///
+///     QkParam *a = qk_param_new("a");
+///     char* str = qk_param_to_string(a);
+///     printf(str);
```

Comment:
It would be nice to include freeing both the param and, more importantly, freeing the string here via `qk_str_free` to teach users they how to free strings obtained from Qiskit's C API